File: app/src/main/java/sergirex/portadasperiodicos/PortadaDetalleFragment.java
package sergirex.portadasperiodicos;

import android.app.ProgressDialog;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.squareup.picasso.Picasso;

import java.util.Objects;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link PortadaDetalleFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class PortadaDetalleFragment extends Fragment {

    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_TITLE = "title";
    private static final String ARG_URL = "url";

    // TODO: Rename and change types of parameters
    private String title;
    private String url;

    public PortadaDetalleFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param title Parameter 1.
     * @param url Parameter 2.
     * @return A new instance of fragment PortadaDetalleFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static PortadaDetalleFragment newInstance(String title, String url) {
        PortadaDetalleFragment fragment = new PortadaDetalleFragment();
        Bundle args = new Bundle();
        args.putString(ARG_TITLE, title);
        args.putString(ARG_URL, url);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            title = getArguments().getString(ARG_TITLE);
            url = getArguments().getString(ARG_URL);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView = inflater.inflate(R.layout.fragment_portada_detalle, container, false);

        ImageView imageView = rootView.findViewById(R.id.imagen_extendida);
        ProgressDialog pd = new ProgressDialog(rootView.getContext());
        pd.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
        pd.setCancelable(false);
        pd.setIndeterminate(true);
        pd.show();
        Picasso.get().load(url)
                .into(imageView, new com.squareup.picasso.Callback() {
                    @Override
                    public void onSuccess() {
                        pd.cancel();
                        pd.dismiss();
                    }

                    @Override
                    public void onError(Exception e) {
                        pd.cancel();
                        pd.dismiss();
                        Picasso.get().load(url)
                                .into(imageView);
                    }
                });

        return rootView;
    }

    @Override
    public void onResume() {
        super.onResume();
        String title = this.title.replace("_", " ");
        title = title.substring(0, 1).toUpperCase() + title.substring(1);
        Objects.requireNonNull(((AppCompatActivity) requireActivity()).getSupportActionBar()).setTitle(title);
    }
}

File: app/src/main/java/sergirex/portadasperiodicos/PortadaDetalleFragment.kt
package sergirex.portadasperiodicos

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.squareup.picasso.Picasso
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import sergirex.portadasperiodicos.databinding.FragmentPortadaDetalleBinding
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

class PortadaDetalleFragment : Fragment() {

    // Use View Binding for safe and efficient view access
    private var _binding: FragmentPortadaDetalleBinding? = null
    private val binding get() = _binding!!

    // Properties to hold fragment arguments
    private var newspaperTitle: String? = null
    private var countryCode: String? = null
    private var initialDate: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Retrieve arguments passed to the fragment
        arguments?.let {
            newspaperTitle = it.getString(ARG_TITLE)
            countryCode = it.getString(ARG_SIGLA_PAIS)
            initialDate = it.getString(ARG_FECHA)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        // Inflate the layout using View Binding
        _binding = FragmentPortadaDetalleBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Launch a coroutine to fetch the cover image in the background
        viewLifecycleOwner.lifecycleScope.launch {
            getCover()
        }
    }

    private suspend fun getCover() {
        // Ensure we have the necessary data to proceed
        if (newspaperTitle == null || countryCode == null || initialDate == null) {
            showError("Información insuficiente para cargar la portada.")
            return
        }

        val formatter = SimpleDateFormat("yyyy/MM/dd", Locale.FRANCE)
        val calendar = Calendar.getInstance()
        try {
            calendar.time = formatter.parse(initialDate!!)!!
        } catch (e: Exception) {
            showError("Fecha inválida.")
            return
        }

        var coverFound = false
        var finalDate: String? = null

        // Try to find the cover, going back up to 20 days
        for (i in 0 until 20) {
            val currentDate = formatter.format(calendar.time)
            val url = "https://img.kiosko.net/$currentDate/$countryCode/$newspaperTitle.jpg"

            // Run the network request on a background thread
            val success = withContext(Dispatchers.IO) {
                loadImageWithPicasso(url)
            }

            if (success) {
                coverFound = true
                finalDate = currentDate
                break // Exit the loop as soon as the cover is found
            } else {
                // If not found, go to the previous day
                calendar.add(Calendar.DATE, -1)
            }
        }

        // Update the UI on the main thread
        binding.loadingProgressBar.visibility = View.GONE
        if (coverFound && finalDate != null) {
            showDateIfNotToday(finalDate)
            binding.imagenExtendida.visibility = View.VISIBLE
        } else {
            showError("No se pudo encontrar ninguna portada.")
        }
    }

    // A helper suspend function to wrap Picasso's callback in a coroutine
    private suspend fun loadImageWithPicasso(url: String): Boolean = suspendCoroutine { continuation ->
        Picasso.get().load(url).into(binding.imagenExtendida, object : com.squareup.picasso.Callback {
            override fun onSuccess() {
                continuation.resume(true) // Resume coroutine with success
            }

            override fun onError(e: Exception?) {
                continuation.resume(false) // Resume coroutine with failure
            }
        })
    }

    private fun showDateIfNotToday(coverDate: String) {
        val formatter = SimpleDateFormat("yyyy/MM/dd", Locale.FRANCE)
        val todayCalendar = Calendar.getInstance()
        if (todayCalendar.get(Calendar.HOUR_OF_DAY) < 6) {
            todayCalendar.add(Calendar.DATE, -1)
        }
        val todayDate = formatter.format(todayCalendar.time)

        if (coverDate != todayDate) {
            val dateStr = coverDate.split("/")
            binding.fechaPortada.text = "${dateStr[2]}/${dateStr[1]}/${dateStr[0]}"
            binding.fechaPortada.visibility = View.VISIBLE
        }
    }

    private fun showError(message: String) {
        binding.loadingProgressBar.visibility = View.GONE
        Toast.makeText(requireContext(), message, Toast.LENGTH_LONG).show()
        // Optionally, navigate back or show an error image
        parentFragmentManager.popBackStack()
    }

    override fun onResume() {
        super.onResume()
        // Set the ActionBar title
        val formattedTitle = newspaperTitle?.replace("_", " ")?.replaceFirstChar { it.uppercase() }
        (activity as? AppCompatActivity)?.supportActionBar?.title = formattedTitle
    }

    override fun onDestroyView() {
        super.onDestroyView()
        // Avoid memory leaks by nullifying the binding reference
        _binding = null
    }

    // Companion object to provide a factory method for creating the fragment
    companion object {
        private const val ARG_TITLE = "title"
        private const val ARG_SIGLA_PAIS = "sigla_pais"
        private const val ARG_FECHA = "fecha"

        @JvmStatic
        fun newInstance(title: String, siglaPais: String, fecha: String) =
            PortadaDetalleFragment().apply {
                arguments = Bundle().apply {
                    putString(ARG_TITLE, title)
                    putString(ARG_SIGLA_PAIS, siglaPais)
                    putString(ARG_FECHA, fecha)
                }
            }
    }
}